leaderelection/config.go

```diff
@@ -117,13 +117,26 @@ func defaultConfig() *Config {
 
 // ComponentConfig represents the leader election config for a single component.
 type ComponentConfig struct {
-	Component     string
-	LeaderElect   bool
-	Buckets       uint32
+	Component   string
+	LeaderElect bool
+	Buckets     uint32
+
+	// The following section is for standard leader election mode only.
 	ResourceLock  string
 	LeaseDuration time.Duration
 	RenewDeadline time.Duration
 	RetryPeriod   time.Duration
+
+	// The following section is for StatefulSet ordinal assignment mode only.
+	StatefulSet StatefulSetConfig
```

Comment:
tl;dr I don't think this should be in ComponentConfig

I think that we should pass this configuration in as an additional parameter to `WithStatefulSetLeaderElectorBuilder`.

I also think that we should have a function that wraps that and the standard version and attempts to extract the `StatefulSetConfig` from the env to set up the appropriate version.  happy to jump on a call if you think it'd be helpful.

Reply:
Also I think we should update `sharedmain` to use the version that does detection.

Reply:
Passed `StatefulSetConfig` in as an additional parameter to `WithStatefulSetLeaderElectorBuilder`. Didn't get the idea about `a function that wraps that and the standard version and attempts to extract the StatefulSetConfig from the env to set up the appropriate version`.

Reply:
Basically what I was suggesting was something like:
```
func WithLeaderElectorBuilder(ctx context.Context) context.Context {
  // This should perhaps be envconfig.Process()
  if os.Getenv("CONTROLLER_ORDINAL") {
    return WithStatefulSetLeaderElectorBuilder(ctx)
  }
  return WithStandardLeaderElectorBuilder(ctx)
}
```